addons/mass_mailing/controllers/main.py

```diff
@@ -148,101 +358,121 @@ def full_url_redirect(self, code, mailing_trace_id, **post):
     def mailing_report_deactivate(self, token, user_id):
         if not token or not user_id:
-            raise NotFound()
-        user_id = int(user_id)
-        correct_token = consteq(token, request.env['mailing.mailing']._generate_mailing_report_token(user_id))
-        user = request.env['res.users'].sudo().browse(user_id)
-        if correct_token and user.has_group('mass_mailing.group_mass_mailing_user'):
-            request.env['ir.config_parameter'].sudo().set_param('mass_mailing.mass_mailing_reports', False)
-            if user.has_group('base.group_system'):
-                menu_id = request.env.ref('mass_mailing.menu_mass_mailing_global_settings').id
-                return request.render('mass_mailing.mailing_report_deactivated', {'menu_id': menu_id})
-            return request.render('mass_mailing.mailing_report_deactivated')
-        raise NotFound()
+            raise BadRequest()
+        user = request.env['res.users'].sudo().browse(int(user_id)).exists()
+        if not user or not user.has_group('mass_mailing.group_mass_mailing_user') or \
+           not consteq(token, request.env['mailing.mailing']._generate_mailing_report_token(user.id)):
+            raise Unauthorized()
+
+        request.env['ir.config_parameter'].sudo().set_param('mass_mailing.mass_mailing_reports', False)
+        render_vals = {}
+        if user.has_group('base.group_system'):
+            render_vals = {'menu_id': request.env.ref('mass_mailing.menu_mass_mailing_global_settings').id}
+        return request.render('mass_mailing.mailing_report_deactivated', render_vals)
 
     @http.route(['/mailing/<int:mailing_id>/view'], type='http', website=True, auth='public')
-    def mailing_view_in_browser(self, mailing_id, email=None, res_id=None, token=""):
-        mailing = request.env['mailing.mailing'].sudo().browse(mailing_id)
-        if mailing.exists():
-            res_id = int(res_id) if res_id else False
-            if not self._check_mailing_email_token(mailing_id, res_id, email, str(token)) and not request.env.user.has_group('mass_mailing.group_mass_mailing_user'):
-                raise exceptions.AccessDenied()
-
-            # do not force lang, will simply use user context
-            html_markupsafe = mailing._render_field(
-                'body_html',
-                [res_id],
-                compute_lang=False,
-                options={'post_process': False}
-            )[res_id]
-            # Update generic URLs (without parameters) to final ones
-            html_markupsafe = html_markupsafe.replace('/unsubscribe_from_list',
-                                                      mailing._get_unsubscribe_url(email, res_id))
-
-            return request.render('mass_mailing.mailing_view', {
-                    'body': html_markupsafe,
-                })
-
-        return request.redirect('/web')
+    def mailing_view_in_browser(self, mailing_id, email=None, document_id=None, hash_token=""):
```

Comment:
when I test on the runbot I have an error when going on this link, as public user
(unsubscribe work, this link work only if I'm logged as mailing admin, not as public / portal user)

because of that check in `_check_mailing_email_token`
```py
if mailing_id and not request.env.user.has_group('mass_mailing.group_mass_mailing_user'):
    raise BadRequest()
```

But in `_prepare_outgoing_list`, we replace all link "/view" with that URL so public / portal user can technically receive that link

Reply:
Indeed, /view was not updated, good catch :) I updated it, and also added a backward compatible check on /view routes. That way people using old links can still "view in desktop" mailings sent before the code update.